Document/0x05c-Reverse-Engineering-and-Tampering.md

```diff
@@ -1014,6 +1014,34 @@ $ frida-trace -p 1372 -a "libjpeg.so!0x4793c"
 
 To learn more about all options for advanced usage, check the [documentation in the official Frida website](https://frida.re/docs/frida-trace/ "documentation").
 
+##### JNI Tracing
+As detailed in the section on [Static Analysis](#static-analysis), the first argument passed to every JNI function is a JNI interface pointer. This pointer contains a table of functions that allows native code to access the Android Runtime. Spotting this pattern can help with understanding the type of functionality a library is performing, such as where it creates new strings or calls Java methods.
+
+[jnitrace](https://github.com/chame1eon/jnitrace "jnitrace") is a Frida based tool that traces usage of the methods in the JNI interface pointer by a library, including the arguments and return values. It is heavily influenced by `frida-trace` and aims to provide similar functionality and output.
+
+Like `frida-trace`, `jnitrace` can be installed via pip. However, `jnitrace` is only available for python3.
+
+```bash
+$ pip install jnitrace
+```
+
+The simplest way to run `jnitrace` is as follows:
+
+```bash
+$ jnitrace -l libnative-lib.so sg.vantagepoint.helloworldjni
+```
+
+<img src="Images/Chapters/0x05c/jnitrace_trace_helloworldjni.png" width="500px"/>
```

Comment:
I'm not able to visualize the image when going to "View file". could you please verify that? maybe it has something to be with the extension?

Reply:
Good spot. The file extension was capatilized, so worked on Windows but not elsewhere. I have changed the filename and made sure that the extension is lowercase.